akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -260,8 +260,9 @@ private[http] final class GracefulTerminatorStage(settings: ServerSettings)
       })
 
       def installTerminationHandlers(deadline: Deadline): Unit = {
-        // when no inflight requests, complete stage right away
-        if (!pendingUserHandlerResponse) completeStage()
+        // when no inflight requests, fail stage right away, could probably be a complete
+        // when https://github.com/akka/akka-http/issues/3209 is fixed
+        if (!pendingUserHandlerResponse) failStage(new ServerTerminationDeadlineReached)
```

Comment:
Another part of the puzzle in #3209: when `!pendingUserHandlerResponse` but there's an ongoing response stream, we completed before which is just to tame to have any effect. Now we bring the connection down immediately even before the hard deadline in such a case.